### ISSUE TITLE
parameter only used in recursion false positive

### ISSUE THREAD

**Issue by hellow554** [rust-lang/rust-clippy#8629]
### Summary

The lint triggers even when there's no recursion and it is "used"

### Lint Name

only_used_in_recursion

### Reproducer

I tried this code:

```rust
pub fn k(num_chunks: u64)
{
    let num_bytes = 98989;
    let _chunk_size = num_bytes * num_chunks;
}
```

I saw this happen:

```
warning: parameter is only used in recursion
 --> src/main.rs:1:10
  |
1 | pub fn k(num_chunks: u64)
  |          ^^^^^^^^^^ help: if this is intentional, prefix with an underscore: `_num_chunks`
  |
  = note: `#[warn(clippy::only_used_in_recursion)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#only_used_in_recursion
```

### Version

```text
rustc 1.61.0-nightly (6af09d250 2022-04-03)
binary: rustc
commit-hash: 6af09d2505f38e4f1df291df56d497fb2ad935ed
commit-date: 2022-04-03
host: x86_64-unknown-linux-gnu
release: 1.61.0-nightly
LLVM version: 14.0.0
```

**Comment by tony84727** [comment]
Maybe another similar false-positive example:
```rust
fn doing_thing(x: i32, y: i32) -> i32 {
    let mut diff = x ^ y;
    let mut distance = 0;
    for _ in 0..32 {
        if diff & 1 > 0 {
            distance += 1;
        }
        diff >>= 1;
    }
    distance
}
fn main() {
    println!("{}", doing_thing(100, 100));
}
```
Clippy returned:
```
warning: parameter is only used in recursion
 --> src/main.rs:1:16
  |
1 | fn doing_thing(x: i32, y: i32) -> i32 {
  |                ^ help: if this is intentional, prefix with an underscore: `_x`
  |
  = note: `#[warn(clippy::only_used_in_recursion)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#only_used_in_recursion

warning: parameter is only used in recursion
 --> src/main.rs:1:24
  |
1 | fn doing_thing(x: i32, y: i32) -> i32 {
  |                        ^ help: if this is intentional, prefix with an underscore: `_y`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#only_used_in_recursion

```

**Comment by hellow554** [comment]
@tony84727 this is somewhat the same pattern as mine:

```rust
pub fn doing_thing(x: i32, y: i32) {
    let _diff = x ^ y;
}
```

```text
warning: parameter is only used in recursion
 --> src/lib.rs:1:20
  |
1 | pub fn doing_thing(x: i32, y: i32) {
  |                    ^ help: if this is intentional, prefix with an underscore: `_x`
  |
  = note: `#[warn(clippy::only_used_in_recursion)]` on by default
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#only_used_in_recursion

warning: parameter is only used in recursion
 --> src/lib.rs:1:28
  |
1 | pub fn doing_thing(x: i32, y: i32) {
  |                            ^ help: if this is intentional, prefix with an underscore: `_y`
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#only_used_in_recursion
```

But yeah, that should be fixed to. I try to investigate this a little bit further.

**Comment by hellow554** [comment]
@buttercrab can you help us here? :)

**Comment by buttercrab** [comment]
@hellow554 I'll change the algorithm as soon as I can.

**Comment by hellow554** [comment]
No need to hurry! Take your time. I just wanted to know, if you know what could cause this. Sorry for the scare

**Comment by danieleades** [comment]
this is causing a false positive in [`rust-base64`](https://github.com/marshallpierce/rust-base64) - https://github.com/marshallpierce/rust-base64/issues/188
